FastSimulation/Tracking/src/SeedFinderSelector.cc

```diff
@@ -119,8 +153,108 @@ bool SeedFinderSelector::pass(const std::vector<const FastTrackerRecHit *>& hits
+	SeedingLayerSetsHits::SeedingLayerSet pairCandidate;
+        //hit 1                                                                                                                                                                
+        if( (hits[i]->det()->geographicalId()).subdetId() == PixelSubdetector::PixelBarrel){
+          layerNo = (*trackerTopology.product()).pxbLayer(hits[i]->det()->geographicalId());
+          IsPixB = true;
+        }
+        else if ((hits[i]->det()->geographicalId()).subdetId() == PixelSubdetector::PixelEndcap){
+          layerNo = (*trackerTopology.product()).pxfDisk(hits[i]->det()->geographicalId());
+          side = (*trackerTopology.product()).pxfSide(hits[i]->det()->geographicalId())==1 ? "_neg" : "_pos";
+          IsPixB = false;
+        }
+        hitlayer[0] = LayerName(layerNo, side, IsPixB);
+        //hit 2                                                                                                                                                                
+        if( (hits[i+1]->det()->geographicalId()).subdetId() == PixelSubdetector::PixelBarrel){
+          layerNo = (*trackerTopology.product()).pxbLayer(hits[i+1]->det()->geographicalId());
+          IsPixB = true;
+        }
+        else if ((hits[i+1]->det()->geographicalId()).subdetId() == PixelSubdetector::PixelEndcap){
+          layerNo = (*trackerTopology.product()).pxfDisk(hits[i+1]->det()->geographicalId());
+          side = (*trackerTopology.product()).pxfSide(hits[i+1]->det()->geographicalId())==1 ? "_neg" : "_pos";
+          IsPixB = false;
+        }
+	hitlayer[1] = LayerName(layerNo, side, IsPixB);
+        for(SeedingLayerSetsHits::SeedingLayerSet ls : *seedingLayer){
+          for(const auto p : layerPairs_){
+            pairCandidate = ls.slice(p,p+2);
+	    std::string layerPair[2] = {};
+            int i=0;
+            for(auto layerSet : pairCandidate){
+              layerPair[i] = layerSet.name();
+              i++;
+            }
+            if((layerPair[0] == hitlayer[0] && layerPair[1] == hitlayer[1]))
+              break;
+          }
+        }
+	const DetLayer * fLayer = measurementTracker_->geometricSearchTracker()->detLayer(hits[i]->det()->geographicalId());
+        const DetLayer * sLayer = measurementTracker_->geometricSearchTracker()->detLayer(hits[i+1]->det()->geographicalId());
+	std::vector<BaseTrackerRecHit const *> fHits(1,static_cast<const BaseTrackerRecHit*>(hits[i]));
+	std::vector<BaseTrackerRecHit const *> sHits(1,static_cast<const BaseTrackerRecHit*>(hits[i+1]));
+        const RecHitsSortedInPhi firsthm(fHits, trackingRegion_->origin(), fLayer);
+        const RecHitsSortedInPhi secondhm(sHits, trackingRegion_->origin(), sLayer);
+        HitDoublets res(firsthm,secondhm);
+	HitPairGeneratorFromLayerPair::doublets(*trackingRegion_,*fLayer,*sLayer,firsthm,secondhm,*eventSetup_,0,res);
+        impl_->produce(layers, pairCandidate, *trackingRegion_, std::move(res), *event_);
```

Comment:
What you should do here instead is something along
```cpp
// outside of the for-loop over hits
IntermediateHitDoublets ihd(seedingLayer);
auto filler = ihd.beginRegion(*trackingRegion_);

  // here
  filler.addDoublets(pairCandidate, std::move(res));

// after the loop
std::vector<OrderedHitSeeds> quadrupletResult;
CAHitQuadGenerator_->hitNtuplets(ihd, quadrupletResult, *eventSetup_, *seedingLayer);
```

Then I'm not sure if the hits would have to be added to the `SeedingLayerSetsHits` as well. I had imagined "yes", but I don't see now a reason why they would have to, as you pass the hits explicitly to `HitPairGeneratorFromLayerPair`, and CA operates purely on the doublets.

Reply:
I did the IHD implementation this way and it seems to work with doublets being added properly to filler. The hitNtuplets() also works, control actually goes inside the function. However, quadrupletResult is always 'false'. I don't know what's going wrong but I think I know where the problem starts. So, inside this loop 
https://github.com/cms-sw/cmssw/blob/master/RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc#L24
with rootVertex=0, it fails to enter this next loop
https://github.com/cms-sw/cmssw/blob/master/RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc#L29
and then whatever happens, finally numberOfFoundQuadruplets=0 always at line 
https://github.com/cms-sw/cmssw/blob/master/RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc#L229

Reply:
Also, FastSim RecHits have to be added to SeedingLayerSetsHits as well since it is used by the CA class to create the graph structure here,
https://github.com/cms-sw/cmssw/blob/master/RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc#L93
I have managed to do it properly this time and it works.

Reply:
> However, quadrupletResult is always 'false'. I don't know what's going wrong but I think I know where the problem starts.

Could you give me a recipe to reproduce? I couldn't come up what could go wrong just by staring the code for a while so I have to try myself. In the meantime I have commented further #3.